src/memory_pool.rs

```diff
@@ -0,0 +1,181 @@
+/// Wrapper around a DataFusion MemoryPool that records allocations,
+/// deallocations, usage. Note that not all query operators register
+/// themselves with the pool, so this isn't an exact limit or breakdown
+/// of the memory used by the query processing.
+use std::sync::Arc;
+
+use datafusion::execution::memory_pool::{MemoryConsumer, MemoryPool, MemoryReservation};
+use metrics::{counter, describe_gauge, gauge, Gauge};
+
+const ALLOCATIONS: &str = "seafowl_datafusion_memory_pool_allocated_bytes_total";
+const DEALLOCATIONS: &str = "seafowl_datafusion_memory_pool_freed_bytes_total";
+const RESERVED: &str = "seafowl_datafusion_memory_pool_reserved_bytes_current";
+
+struct Metrics {
+    memory_reserved: Gauge,
+}
+
+impl Metrics {
+    pub fn new() -> Self {
+        describe_gauge!(
```

Comment:
Late to the party, but should this be `describe_counter` rather than `describe_gauge`? Ditto in L24

Reply:
Oh, that's actually a good catch - these are used as counters below and not gauges, though for some reason they still show up in the metrics output as `type counter`.

```
$ curl localhost:9191/metrics | grep memo | grep TYPE
# TYPE seafowl_datafusion_memory_pool_allocated_bytes_total counter
# TYPE seafowl_datafusion_memory_pool_freed_bytes_total counter
# TYPE seafowl_datafusion_memory_pool_reserved_bytes_current gauge
```

Cowboy-pushed 2755bb85233b5c6eb5d669246fb54ac9528b7a53 to fix.